### DISCUSSION TITLE
Save original taxa names

### DISCUSSION THREAD

**Discussion by davised** [david-barnett/microViz#88] [Ideas]
Hi David,

Just getting this idea a little more fleshed out on paper.

### Suggestion

Save the original taxa names such that they are accessible even after renaming using `tax_name` or `tax_rename`.

### Rationale

Users of the dada2 -> phyloseq analysis pipeline will be interested in renaming the long ASV sequence names that are output by dada2. The `tax_name` and `tax_rename` functions can help with this end goal. Yet, users may want to access those ASV sequences, or other original identifiers, as a reference between different types of analyses, even as the display names are renamed.

Secondarily, we can make it easy to print the renamed taxon name along with the original taxon name, either as tab-delimited or FASTA format, as appropriate, such that the data can be used in external programs, e.g. phylogenetic tree software.

### Potential implementation

I suggest saving the original taxa names as a slot in the psExtra object. This could happen even prior to any `tax_name` or `tax_rename` calls. Later, after taxa are renamed, a `tax_save` function could be called, along with options about how the output is printed.

### Example code

This is what I've done in the past, as an example:

```r
# Convert names to arbitrary OTU names
names.ps <- as.data.frame(cbind(taxa_names(ps),
                                paste0("ASV", seq(ntaxa(ps)))),
                          stringsAsFactors=F)
colnames(names.ps) <- c('seq', 'asv')
taxa_names(ps) <- names.ps$asv

# write to file
write(paste(paste0(">", names.ps$asv), names.ps$seq, sep="\n"), "ASVs.fasta")
```

What are your thoughts?

Cheers,

Ed

**Comment by david-barnett** [comment]
Thanks Ed! I'll think about this more carefully probably next week, 

Some initial thoughts

- maybe if the main/only use-case is sequences then the refseq slot from standard phyloseq (or some enhancement of it) is a good place to start

- Maybe a useful thing could be making the transfer to and from taxa_names and refseq slot easy? 

- or perhaps the enforced XStringset format might be too restrictive in the refseq slot, which could justify another slot for generic names. 

- Another angle might be using tax_name2rank to store the sequences in the tax_table, maybe even giving a "sequence" rank special behaviour, like validating that all values are unique and a special print method
- That way, aggregation at a higher rank would drop the sequences from that object, but I think that makes sense? 

cheers
David

**Reply by davised** [reply]
Getting back to this discussion.

I think you are on point with your thoughts. 

I'm imagining that the 'easiest' way would be to make another rank in the tax_table as you suggested. That's probably the most visible method to users and relatively understandable.

I like the refseq idea, though, because I think it would be easier to implement the refseq to fasta print function that I suggested. Not sure how the refseq would work in the case of aggregating taxa, while the tax_table method would likely work without too much hassle. Enforcing XStringset isn't a bad thing in my mind if we are trying to enforce sequences only, but maybe we don't need to enforce sequences only.

Thanks for considering this idea. Loving the updates at 0.10.x thus far.

Cheers.
